extends Node3D
## Static player
## Made by Yni, licensed under MIT License.
class_name StaticPlayer

enum CameraMode {ALL, UPPERLOOK, THIRD_PERSON, SIZE}

@export var camera_mode: CameraMode = CameraMode.ALL
@export var current_camera_mode: CameraMode = CameraMode.UPPERLOOK
@export var target_puppet_path: String = ""
var mouse_sensitivity = 0.03125
var prev_x_coordinate: float = 0
var scroll_factor: float = 1.0
var transition: NodePath


const RAY_LENGTH = 512

# Called when the node enters the scene tree for the first time.
func _ready() -> void:
	pass # Replace with function body.

func _input(event: InputEvent) -> void:
	if event is InputEventMouseMotion:
		if Input.is_action_pressed("look"):
			rotate_player(event)
	if event is InputEventScreenDrag:
		rotate_player(event)
	#if event.is_action_pressed("scroll_up"):
		#scroll_factor += 0.125
		#scroll_factor = clamp(scroll_factor, 1.0, 8.0)
		#$Head/Camera3D.fov = 75.0 / scroll_factor
	#if event.is_action_pressed("scroll_down"):
		#scroll_factor -= 0.125
		#scroll_factor = clamp(scroll_factor, 1.0, 8.0)
		#$Head/Camera3D.fov = 75.0 / scroll_factor


# Called every frame. 'delta' is the elapsed time since the previous frame.
func _physics_process(delta: float) -> void:
	#Handle smooth camera transitions
	if transition != null && !transition.is_empty():
		var to_pos: Vector3 = get_node(transition).position
		$Head/Camera3D.position = $Head/Camera3D.position.move_toward(to_pos, 12 * delta)
		if $Head/Camera3D.position.is_equal_approx(to_pos):
			transition = NodePath()
	if Input.is_action_just_pressed("toggle_mode"):
		toggle_mode(current_camera_mode + 1 if current_camera_mode + 1 < CameraMode.SIZE else 1)
	rotate_player_by_key(Vector2i(int(Input.is_action_just_pressed("camera_rotate_right")) - int(Input.is_action_just_pressed("camera_rotate_left")), 0))
	if !target_puppet_path.is_empty():
		if get_node_or_null(target_puppet_path) == null:
			get_tree().root.get_node("Game").finish_game(false, "GAME_OVER_1")
		else:
			if get_node(target_puppet_path).current_health[2] < get_node(target_puppet_path).health[2]:
				$Head/Camera3D/MeshInstance3D.mesh.surface_get_material(0).set_shader_parameter("multiplier", (get_node(target_puppet_path).health[2] - get_node(target_puppet_path).current_health[2]) / get_node(target_puppet_path).health[2])
			get_tree().root.get_node("Game/UI/HealthBar").value = get_node(target_puppet_path).current_health[0]
		# Apply bonus to Y coordinate if current_camera_mode is third person
		if current_camera_mode == CameraMode.THIRD_PERSON:
			global_position = get_node(target_puppet_path).global_position + Vector3(0, 3, 0) + Vector3(0, 0.875, 0)
		else:
			global_position = get_node(target_puppet_path).global_position + Vector3(0, 3, 0)

## Used from Godot Docs
func intersect() -> Dictionary:
	var space_state = get_world_3d().direct_space_state
	var mousepos = get_viewport().get_mouse_position()

	var origin = $Head/Camera3D.project_ray_origin(mousepos)
	var end = origin + $Head/Camera3D.project_ray_normal(mousepos) * RAY_LENGTH
	var query = PhysicsRayQueryParameters3D.create(origin, end)
	
	return space_state.intersect_ray(query)

## Used from Godot Docs
func intersect_shape(intersect_position: Vector3) -> Array[Dictionary]:
	var space_state = get_world_3d().direct_space_state
	var mousepos = get_viewport().get_mouse_position()
	
	var shape_rid = PhysicsServer3D.sphere_shape_create()
	var radius = 3.0
	PhysicsServer3D.shape_set_data(shape_rid, radius)

	var params = PhysicsShapeQueryParameters3D.new()
	params.shape_rid = shape_rid
	params.transform = Transform3D(Basis(), intersect_position)
	params.collision_mask = 10
	
	var result: Array[Dictionary] = space_state.intersect_shape(params, 4)
	
	# Release the shape when done with physics queries.
	PhysicsServer3D.free_rid(shape_rid)
	
	return result

func interact(value: String) -> void:
	if get_node_or_null(target_puppet_path) == null:
		get_tree().root.get_node("Game").finish_game(false, "GAME_OVER_1")
	else:
		match value:
			"Point":
				var result: Dictionary = intersect()
				if result.keys().size() > 0 && !get_node(target_puppet_path).movement_freeze:
					# Shape cast for items
					var shape_result: Array[Dictionary] = intersect_shape(result["position"])
					for s_result in shape_result:
						if s_result.keys().size() > 0:
							if s_result["collider"] is Pickable && !s_result["collider"].freeze && s_result["collider"].global_position.distance_to(get_node(target_puppet_path).global_position) < 4.0:
								get_node(get_tree().root.get_node("Game/StaticPlayer").target_puppet_path + "/UI/Inventory/Inventory").add_item(s_result["collider"].item_id)
								s_result["collider"].queue_free()
								#Use only one item
								break
							if s_result["collider"] is MovableNpc:
								var test = !s_result["collider"].puppet_class.automatic
								if !s_result["collider"].is_player && !s_result["collider"].puppet_class.automatic:
									s_result["collider"].follow_target = target_puppet_path
									if s_result["collider"].wandering:
										s_result["collider"].wandering = false
					# ray cast for moving
					if get_node_or_null(target_puppet_path) == null:
						get_tree().root.get_node("Game").finish_game(false, "GAME_OVER_1")
					else:
						get_node(target_puppet_path).set_target_position(result["position"])

func rotate_player(event: InputEvent):
	# Yni: Necessary to fix annoying bug on Android, when if you rotate screen, player began to move.
	# https://kidscancode.org/godot_recipes/3.x/3d/camera_gimbal/index.html
	rotate_object_local(Vector3.UP, event.relative.x * mouse_sensitivity * 0.05)
	var y_rotation = clamp(event.relative.y, -30, 30)
	$Head.rotate_object_local(Vector3.RIGHT, y_rotation * mouse_sensitivity * 0.05)
	$Head.rotation_degrees.x = clamp($Head.rotation_degrees.x, -90, 0)
	#rotation.y -= event.relative.x * mouse_sensitivity * 0.05
	#rotation.x -= event.relative.y * mouse_sensitivity * 0.05
	#rotation_degrees.y = clamp(rotation_degrees.y, -90, 90)

func rotate_player_by_key(direction: Vector2i):
	var x_dir: float
	var y_dir: float
	match direction:
		Vector2i.UP:
			y_dir = 15
		Vector2i.DOWN:
			y_dir = -15
		Vector2i.LEFT:
			x_dir = -45
		Vector2i.RIGHT:
			x_dir = 45
	# Yni: Necessary to fix annoying bug on Android, when if you rotate screen, player began to move.
	# https://kidscancode.org/godot_recipes/3.x/3d/camera_gimbal/index.html
	rotate_object_local(Vector3.UP, deg_to_rad(x_dir))
	var y_rotation = clamp(y_dir, -30, 30)
	$Head.rotate_object_local(Vector3.RIGHT, deg_to_rad(y_rotation))
	$Head.rotation_degrees.x = clamp($Head.rotation_degrees.x, -90, 0)

func toggle_mode(mode: int):
	if mode == 0 || mode >= CameraMode.SIZE:
		printerr("Cannot specify incompatible mode")
	match mode:
		1:
			if camera_mode == 0 || camera_mode == 1:
				global_position.y = 0
				transition = $Head/UpperLook.get_path()
				current_camera_mode = CameraMode.UPPERLOOK
			else:
				printerr("camera_mode does not allow this mode")
		2:
			if camera_mode == 0 || camera_mode == 2:
				global_position.y = 1.863
				transition = $Head/ThirdPerson.get_path()
				current_camera_mode = CameraMode.THIRD_PERSON
			else:
				printerr("camera_mode does not allow this mode")

func _on_optimizator_body_entered(body: Node3D) -> void:
	if body is MovableNpc:
		body.optimizator_paused = false


func _on_optimizator_body_exited(body: Node3D) -> void:
	if body is MovableNpc:
		body.optimizator_paused = true
